cupy/cuda/compiler.py

```diff
@@ -276,6 +278,93 @@ def _hash_hexdigest(value):
 _hash_length = len(_hash_hexdigest(b''))  # 40 for SHA1
 
 
+class RestrictedUnpickler(pickle.Unpickler):
+    """An unpickler that only allow some safe classes."""
+
+    def find_class(self, module, name):
+        """Only allow a restricted set of classes."""
+        if module == __name__ and name == 'CachedCudaModule':
+            return CachedCudaModule
+        msg = "'{}.{}' is forbidden and cannot be unpickled."
+        raise pickle.UnpicklingError(msg.format(module, name))
+
+    @classmethod
+    def loads(cls, s):
+        """Helper function analogous to pickle.loads()."""
+        return cls(io.BytesIO(s)).load()
+
+
+class CachedCudaModule:
+    __slots__ = ['backend', 'cubin', 'mapping']
+
+    _pickle_protocol = 4  # first introduced in Python 3.4
+
+    def __init__(self, backend, cubin, mapping):
+        self.backend = backend
+        self.cubin = cubin
+        self.mapping = mapping
+
+    @classmethod
+    def dump_module(cls, path, backend, cubin, mapping):
+        """Build and dumps a CachedCudaModule object and checksum to path."""
+        python_obj = cls(backend, cubin, mapping)
+
+        data_bytes = pickle.dumps(python_obj, protocol=cls._pickle_protocol)
+        data_hash = _hash_hexdigest(data_bytes).encode('ascii')
+        assert len(data_hash) == _hash_length
+
+        directory = os.path.dirname(path)
+        if not os.path.isdir(directory):
+            os.makedirs(directory, exist_ok=True)
+
+        with open(path, 'wb') as f:
+            f.write(data_hash)
+            f.write(data_bytes)
```

Comment:
This needs to follow the original approach (write to a temporary location and then move). The rationale there was that in HPC environments there could be thousands of processes writing to the same file (if they all need the same kernel). Writing directly to the destination would significantly increases the chances of file corruption, effectively leading to cache invalidation.

In fact, to support cuFFT callbacks we've adopted atomic moves using `os.rename()` (`os.replace()` is also OK), but it's only atomic on Linux. @kmaehashi maybe we can use it here too (once the temporary location is restored as an extra safety layer)?

Reply:
Use of `os.replace()` (but not `os.rename()`) sounds good to me. xref #4926